### PR TITLE
.github/scripts/auto-backport.py: validate backport candidate with `Fixes` prefix

### PR DESCRIPTION
**Pull request by yaronkaikov**
Adding `Fixes` validation to a PR when backport labels were added. When the auto backport process triggers (after promotion), we will ensure each PR with backport/x.y label also has in the PR body a `Fixes` reference to an issue

Fixes: https://github.com/scylladb/scylladb/issues/20021

**Backport validation improvement, no need for backport**

### CODE REVIEW

File: scripts/pull_github_pr.sh [hunk 1]
```diff
@@ -115,12 +115,23 @@ PR_TITLE=$(jq -r .title <<< $PR_DATA)
...
 PR_LOGIN=$(jq -r .head.user.login <<< $PR_DATA)
+PR_LABELS=$(jq -r .labels[].name <<< "$PR_DATA")
+PR_BODY=$(jq -r .body <<< "$PR_DATA")
 echo -n "Fetching full name of author $PR_LOGIN... "
 USER_NAME=$(curl -s "https://api.github.com/users/$PR_LOGIN" | jq -r .name)
 echo "$USER_NAME"
 
 git fetch "$REMOTE" pull/$PR_NUM/head
 
+if [[ "$FORCE" != "--skip-validation" ]]; then
```
Comment (denesb) [inline scripts/pull_github_pr.sh:126]: This means we cannot both skip the jenkins check *and* the validation.
I think it is worth adding a separate flag for this (and also making these flags truly non-positional).

Comment (yaronkaikov) [reply scripts/pull_github_pr.sh:126]: will fix it

**Comment by yaronkaikov** [comment]
Validate the script locally using https://github.com/scylladb/scylladb/pull/21558 (has backport labels but no ref to an issue)

The first attempt failed due to unstable CI and missing ref
```
[yaronkaikov@london]~/git/scylladb (next)$ ./scripts/pull_github_pr.sh 21558 
WARNING: https://jenkins.scylladb.com/job/scylla-master/job/next/lastCompletedBuild is not stable, To override run pull_github_pr.sh with --ignore-ci
Fetching info on PR #21558... 

WARNING: 21558 is a candidate for backport but doesn't contain any Fixes reference to an issue. To override run pull_github_pr.sh with --ignore-ref-check
```

Second attempt with `--ignore-ci` flag
```
[yaronkaikov@london]~/git/scylladb (next)$ ./scripts/pull_github_pr.sh 21558 --ignore-ci
Fetching info on PR #21558... 

WARNING: 21558 is a candidate for backport but doesn't contain any Fixes reference to an issue. To override run pull_github_pr.sh with --ignore-ref-check
```

Third attempt with both flags `--ignore-ci` and `--ignore-ref-check`
```
[yaronkaikov@london]~/git/scylladb (next)$ ./scripts/pull_github_pr.sh 21558 --ignore-ci --ignore-ref-check
Fetching info on PR #21558... 
    utils: UUID: make get_time_UUID() respect clock offset in tests
Fetching full name of author tgrabiec... Tomasz Grabiec
From github.com:scylladb/scylladb
 * branch                  refs/pull/21558/head -> FETCH_HEAD
Auto-merging configure.py
[next 5de178171f] utils: UUID: Make get_time_UUID() respect the clock offset
 Author: Tomasz Grabiec <tgrabiec@scylladb.com>
 Date: Tue Nov 12 19:25:56 2024 +0100
 2 files changed, 3 insertions(+), 3 deletions(-)
[next 2d8bed4283] utils: UUID: Make get_time_UUID() respect the clock offset
 Author: Tomasz Grabiec <tgrabiec@scylladb.com>
 Date: Tue Nov 12 19:25:56 2024 +0100
 2 files changed, 3 insertions(+), 3 deletions(-)
[next 2d8bed4283] utils: UUID: Make get_time_UUID() respect the clock offset
 Author: Tomasz Grabiec <tgrabiec@scylladb.com>
 Date: Tue Nov 12 19:25:56 2024 +0100
 2 files changed, 3 insertions(+), 3 deletions(-)

```

File: scripts/pull_github_pr.sh [hunk 2]
```diff
@@ -28,11 +61,6 @@ if [[ -z "$JENKINS_USERNAME" || -z "$JENKINS_API_TOKEN" ]]; then
   exit 1
 fi
 
-if [ -z "$1" ]; then
-	echo Please provide a github pull request number
-	exit 1
-fi
```
Comment (denesb) [inline scripts/pull_github_pr.sh:34]: We still need this check, you should check `$PR_NUM` now.

File: scripts/pull_github_pr.sh [hunk 3]
```diff
@@ -11,10 +11,43 @@ set -e
...
 jenkins_url="https://jenkins.scylladb.com"
-FORCE=$2
 ORANGE='\033[0;33m'
 NC='\033[0m'
 
+if [ $# -lt 1 ]; then
+  echo Please provide a github pull request number
+fi
+
+PR_NUM=$1
```
Comment (denesb) [inline scripts/pull_github_pr.sh:21]: It is strange to have to pass the flags after the PR number. Can we extract it in the loop below (write it in the else case).

File: scripts/pull_github_pr.sh [hunk 4]
```diff
@@ -11,10 +11,43 @@ set -e
...
+while true; do
+  case "$1" in
+        -i|--ignore-ci)
+            IGNORE_CI=true
+            shift
+            ;;
+        -c|--ignore-ref-check)
+            IGNORE_REF_CHECK=true
+            shift
+            ;;
```
Comment (denesb) [inline scripts/pull_github_pr.sh:42]: I think we should keep `--force`:
```
-f|--force)
    IGNORE_CI=true
    IGNORE_REF_CHECK=true
    shift
    ;;
```

**Review by denesb — CHANGES_REQUESTED** [review]

File: scripts/pull_github_pr.sh (hunk 3 above)
Comment (yaronkaikov) [reply scripts/pull_github_pr.sh]: Fixed it, now you can add the flags in any order

File: scripts/pull_github_pr.sh (hunk 4 above)
Comment (yaronkaikov) [reply scripts/pull_github_pr.sh]: Added

File: scripts/pull_github_pr.sh (hunk 2 above)
Comment (yaronkaikov) [reply scripts/pull_github_pr.sh:34]: Added

**Review by denesb — APPROVED** [review]

**Comment by denesb** [comment]
Can you please re-test just to make sure no regressions slipped in?

**Comment by yaronkaikov** [comment]
> Can you please re-test just to make sure no regressions slipped in?

Another validation

PR without ref
```
[yaronkaikov@london]~/git/scylladb (next)$ ./scripts/pull_github_pr.sh 21386
https://jenkins.scylladb.com/job/scylla-master/job/next/lastCompletedBuild is stable
Fetching info on PR #21386... 

WARNING: 21386 is a candidate for backport but doesn't contain any Fixes reference to an issue. To override run pull_github_pr.sh with --ignore-ref-check
```

Then added `--ignore-ref-check`
```
[yaronkaikov@london]~/git/scylladb (next)$ ./scripts/pull_github_pr.sh --ignore-ref-check 21386
https://jenkins.scylladb.com/job/scylla-master/job/next/lastCompletedBuild is stable
Fetching info on PR #21386... 
    docs: explain task status retention and one-time query behavior
Fetching full name of author tchaikov... Kefu Chai
remote: Enumerating objects: 7, done.
remote: Counting objects: 100% (6/6), done.
remote: Total 7 (delta 6), reused 6 (delta 6), pack-reused 1 (from 1)
Unpacking objects: 100% (7/7), 1.23 KiB | 125.00 KiB/s, done.
From github.com:scylladb/scylladb
 * branch                  refs/pull/21386/head -> FETCH_HEAD
[next e6e1abd1ef] docs: explain task status retention and one-time query behavior
 Author: Kefu Chai <kefu.chai@scylladb.com>
 Date: Thu Oct 31 13:09:30 2024 +0800
 1 file changed, 10 insertions(+)
[next d6b2205ed7] docs: explain task status retention and one-time query behavior
 Author: Kefu Chai <kefu.chai@scylladb.com>
 Date: Thu Oct 31 13:09:30 2024 +0800
 1 file changed, 10 insertions(+)
[next d6b2205ed7] docs: explain task status retention and one-time query behavior
 Author: Kefu Chai <kefu.chai@scylladb.com>
 Date: Thu Oct 31 13:09:30 2024 +0800
 1 file changed, 10 insertions(+)

WARNING: Some of the tests that ran for this PR were not completed successfully,
 please make sure all tests are done successfully before merge this PR.
```

Also with the same flag in different order
```
[yaronkaikov@london]~/git/scylladb (next)$ ./scripts/pull_github_pr.sh 21386 --ignore-ref-check
https://jenkins.scylladb.com/job/scylla-master/job/next/lastCompletedBuild is stable
Fetching info on PR #21386... 
    docs: explain task status retention and one-time query behavior
Fetching full name of author tchaikov... Kefu Chai
From github.com:scylladb/scylladb
 * branch                  refs/pull/21386/head -> FETCH_HEAD
On branch next
Your branch is ahead of 'upstream/next' by 2 commits.
  (use "git push" to publish your local commits)

You are currently cherry-picking commit 94c5b063fd.
```
Also verified the `--force` is working

**Comment by denesb** [comment]
Queued despite failing CI as the PR only touches backport automation. @yaronkaikov please annotate CI failures.

**Comment by yaronkaikov** [comment]
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13168/) ✅ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/220/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1687/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4763/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12786/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13168/)
> 
> #### Failed Tests (1/35659):
> * [test_random_failures[stop_before_becoming_raft_voter-kill_coordinator_node].debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13168/testReport/junit/%28root%29/test_random_failures/Tests___Unit_Tests___test_random_failures_stop_before_becoming_raft_voter_kill_coordinator_node__debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures%5Bstop_before_becoming_raft_voter-kill_coordinator_node%5D.debug.2)
> 
> #### Build Details:
> * Duration: 3 hr 35 min
> * Builder: spider3.cloudius-systems.com

https://github.com/scylladb/scylladb/issues/21534

**Comment by yaronkaikov** [comment]
@avikivity I don't understand why did you revert this,
Can you please share some details about the failure you saw?

@denesb FYI

**Comment by yaronkaikov** [comment]
@denesb re-applying this patch without the changes in `pull_github_pr.sh`

**Pull request closed without merging**